src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
+        return normalized_topic_prefs
+
+
+###################### text generation part
+
+client = OpenAI(
+    api_key="PUT KEY HERE",
+)
+
+
+def gpt_generate(system_prompt, content_prompt):
+    message = [{"role": "system", "content": system_prompt}, {"role": "user", "content": content_prompt}]
+    temperature = 0.2
+    max_tokens = 512
+    frequency_penalty = 0.0
+
+    chat_completion = client.chat.completions.create(
+        messages=message,
+        temperature=temperature,
+        max_tokens=max_tokens,
+        frequency_penalty=frequency_penalty,
+        model="gpt-4o-mini",
+    )
+    return chat_completion.choices[0].message.content
+
+
+def news_factor(headline, slug):
+    news = f"headline: {headline}\nslug: {slug}"
+    system_prompt = (
+        "Based on the given news information, summarize what topics that the news is related to."
+        "Each news article is related to 1-3 topics and each topic shuold not exceed five words."
+    )
+    return gpt_generate(system_prompt, news)
+
+
+# TODO: check backward or forward for past k articles
+def user_interest_generate(past_articles: Article, past_k: int):
+    past_article_infor = []
+    past_count = 0
+    for article in past_articles.articles:
+        past_count += 1
+        if past_count == past_k:
+            break
+        temp_article = {}
+        temp_article["headline"] = article.headline
+        temp_article["subhead"] = article.subhead
+        temp_article["categories"] = set([mention.entity.name for mention in article.mentions])
+        temp_article["topics"] = news_factor(article.headline, article.subhead)
+        past_article_infor.append(temp_article)
+
+    system_prompt = (
+        "You are asked to describe user interest based on his/her browsed news list."
+        " User interest includes the news [categories] and news [topics]"
+        " (under each [category] that users are interested in."
+    )
+
+    return gpt_generate(system_prompt, f"{past_article_infor}")
+
+
+def generate_narrative(news_list, user_interests):
+    system_prompt = (
+        "You are a personalized text generator."
+        " First, i will provide you with a news list that"
+        " includes both the [main news] and [related news]."
+        " Second, I will provide you with user interests,"
+        " including the [categories] of news that the user is interested in."
+        " Based on the input news list and user interests,"
+        " you are required to generate a personalized news summary centered around the [main news]."
+    )
+
+    input_prompt = "News List: \n" + f"{news_list}" + "User Interest: \n" + user_interests
+    return gpt_generate(system_prompt, input_prompt)
+
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
+
+
+def related_article(current_news_slug, current_topics, clicked_articles, k):
+    # TODO: do we need to consider the cold start?
+    target_embedding = model.encode([current_news_slug])[0]
+
+    subject_candidate_headlines = []
```

Comment:
From the code below, I think this is actually a list of subheads? 
```suggestion
    subject_candidate_subheads = []
```